### ISSUE TITLE
Cannot read property 'fragmentedOperation' of null

### ISSUE THREAD

**Issue by toejough** [websockets/ws#366]
Recently, I got this error message on a test teardown for a project using ws:

```
TypeError: Cannot read property 'fragmentedOperation' of null
    at Receiver.endPacket (node_modules/ws/lib/Receiver.js:222:18)
    at Receiver.opcodes.1.finish (node_modules/ws/lib/Receiver.js:394:12)
    at Receiver.<anonymous> (node_modules/ws/lib/Receiver.js:372:33)
    at Receiver.add (node_modules/ws/lib/Receiver.js:87:24)
    at Socket.firstHandler node_modules/ws/lib/WebSocket.js:663:22)
    at Socket.emit (events.js:95:17)
    at Socket.<anonymous> (_stream_readable.js:764:14)
    at Socket.emit (events.js:92:17)
    at emitReadable_ (_stream_readable.js:426:10)
    at emitReadable (_stream_readable.js:422:5)
```

I tracked it down to this general flow of events:
- ws client sends a message, which goes out over the network.
- my application test ends and tears down, closing the local socket (this test only cares about calling the higher-level functions correctly, so the circumstance is contrived, but for the purposes of this bug, let's pretend the network went down)
- the sent data comes in to the server and triggers the `on('data', ...)` event
- ws Receiver.js parses the data, emits the 'message' event
- because node's `emit` actually runs the listeners serially _at the moment 'emit' is called_, my `on('message', ...)` listener is run.
- that listener responds to the message by trying to send data to the client
- the ws Sender.js tries to send the data, but the underlying socket throws an error, which Sender catches and emits an 'error' event for
- again, due to node's immediate execution of listeners, the internal ws `on('error', ...)` listeners are run, which clean up the websocket state
- execution now continues after the internal Receiver.js `this.ontext` call, and proceeds to call `this.endpacket()` (lines 391 and 394 in ws version 0.4.32)
- the first check in `endpacket` is `if (!this.state.fragmentedOperation)`, which explodes because `this.state` is now `null` thanks to the cleanup.

I tried to resolve this by modifying ws to always check state and call `this.error` if it wasn't valid, but `this.error` is also destroyed on cleanup, leaving the user with no way to catch this error condition gracefully, that I could tell.  I mean, as a user, I can catch the initial error event once ws emits it, but ws itself continues on after that and uses the now invalid state.

One way to fix this would be to wrap all `emit` calls in 'process.nextTick', but I'm out of time for tonight.  When I get back to this (maybe in a week?) I'll get started on a pull request for that.

**Comment by ysxlin** [comment]
Hello, I am a node.js beginners. can u tell me how to solve this error?thx

**Comment by toejough** [comment]
The best I've been able to do without patching WS is to make sure that prior to calling ws.send, you check the ws.readyState == ws.OPEN.  If it doesn't, you have to handle that error condition yourself.  You could try:

```js
var client = new ws(url);
client.safeSend = function(data) {
    if (client.readyState == ws.OPEN) {
        client.send(data);
    } else {
        throw new Error("could not send - websocket is not open");
    }
};

// client.send("my message"); // might cause ws to explode, as you have seen
client.safeSend("my message"); // will throw an error you can catch and act on
```

This would throw a new error if you tried to send a message when the websocket wasn't open.  You could emit it instead, (```client.emit('error', new Error(...));```) or log it (```console.error(...);```).

**Comment by luisrudge** [comment]
I'm having this with browsersync+socket.io too :(

**Comment by zachfedor** [comment]
@luisrudge : I'm having this issue with browserSync as well. Running it through a Gulp task, which occasionally spins up fine, but ultimately crashes because of this error. This might not be the right spot for the issue though... Any ideas?

**Comment by luisrudge** [comment]
I had to rollback the browsersync version to `^2.6.4` and it works

**Comment by zachfedor** [comment]
I rolled back to `2.6.0` and it works, too

**Comment by lpinca** [comment]
I think it has been fixed in browserSync 2.7.1, closing.
